### PR TITLE
kvutils: Time transaction acquisition and release.

### PR DESCRIPTION
**Pull request by SamirTalwar-DA**
Part of #5146.

Now we're getting to the bottom of why committing can take so long; we build up a queue.

### Changelog

- **[Ledger Integration Kit]** Metrics for acquiring and releasing transactions.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala [hunk 1]
```diff
@@ -215,23 +223,55 @@ class SubmissionValidator[LogResult](
...
+    }
+  }
+
+  private object Metrics {
     private val prefix = MetricRegistry.name("daml", "kvutils", "submission", "validator")
 
     val openEnvelope: Timer =
       metricRegistry.timer(MetricRegistry.name(prefix, "open_envelope"))
+    val acquireTransaction: Timer =
+      metricRegistry.timer(MetricRegistry.name(prefix, "acquire_transaction"))
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala]: Shouldn't these be `acquire_transaction_lock` and `release_transaction_lock` instead? I.e., acquiring or releasing a transaction doesn't sound right.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala [hunk 2]
```diff
@@ -215,23 +223,55 @@ class SubmissionValidator[LogResult](
...
       case (key, Some(value)) => key -> value
     }
 
-  object Metrics {
+  private final class TimedLedgerStateAccess(delegate: LedgerStateAccess[LogResult])
+      extends LedgerStateAccess[LogResult] {
+    override def inTransaction[T](
+        body: LedgerStateOperations[LogResult] => Future[T]
+    ): Future[T] = {
+      val acquireStopped = new AtomicBoolean(false)
```
Comment (fabiotudone-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala]: @SamirTalwar-DA I guess that if the failure case stops the time, then the inTransaction one won't anymore; on the other hand the inTransaction body might stop it, the transaction might still fail later and then you need to make sure it's not already stopped in the failure halding code (in order not to record a false later time). I think this is the reason why you need to atomic flag but regardless pls. document it in a comment.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala (hunk 1 above)
Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala]: @miklos-da I think @SamirTalwar-DA wanted to call the metric in a way that is independent from the actual acquisition mechanism and I tend to agree.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala (hunk 2 above)
Comment (SamirTalwar-DA) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala]: Yes, good call. I'll also make it a separate timer.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala (hunk 1 above)
Comment (SamirTalwar-DA) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala]: Done. 🙂

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala (hunk 2 above)
Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala]: Even better indeed

**Comment by fabiotudone-da** [comment]
@SamirTalwar-DA interestingly I couldn't find a way to cancel a timer context, it looks like forgetting about it really comes closest

**Comment by SamirTalwar-DA** [comment]
@fabiotudone-da: You don't really need to. Just creating the context has no side effects; it just captures the start time in an object.

**Comment by fabiotudone-da** [comment]
> @fabiotudone-da: You don't really need to. Just creating the context has no side effects; it just captures the start time in an object.

Apart from some GC pressure I suppose

**Comment by SamirTalwar-DA** [comment]
And `while (true) { }` heats up my apartment, but I also wouldn't call that a "side effect". 😜

**Comment by fabiotudone-da** [comment]
> And `while (true) { }` heats up my apartment, but I also wouldn't call that a "side effect".

Yeah but that's bad (i mean GC pressure, not heating your apartment). We are out of luck this time on the JVM though.

**Pull request merged**